tools/validate_reflection.py

```diff
@@ -0,0 +1,713 @@
+                    # this is a chainbase::object, don't need to worry about id_type definition
+                    continue
+        class_struct_num_fields = len(class_struct.fields) if class_struct is not None else None 
+        debug("reflection_name=%s, class field count=%s, reflection field count=%s, ingore count=%s, swap count=%s" % (reflection_name, class_struct_num_fields, len(reflection.fields), len(reflection.ignored), len(reflection.swapped)))
+        assert isinstance(class_struct, ClassStruct), "could not find a %s/%s/%s for %s" % (EmptyScope.class_str, EmptyScope.struct_str, EmptyScope.enum_str, reflection_name)
+        if class_struct.ignore_id:
+            id_field = "id"
+            if id_field not in reflection.ignored and id_field not in reflection.fields:
+                debug("Object ignore_id Adding id to ignored for %s" % (reflection_name))
+                reflection.ignored.append(id_field)
+        else:
+            debug("Object ignore_id NOT adding id to ignored for %s" % (reflection_name))
+        rf_index = 0
+        rf_len = len(reflection.fields)
+
+        processed = []
+        back_swapped = []
+        fwd_swapped = []
+        ignored = []
+        f_index = 0
+        f_len = len(class_struct.fields)
+        while f_index < f_len:
+            field = class_struct.fields[f_index]
+            reflect_field = reflection.fields[rf_index] if rf_index < rf_len else None
+            processed.append(field)
+            debug("\nfield=%s reflect_field=%s" % (field, reflect_field))
+            #debug("start rf_index=%s, rf_len=%s, f_index=%s, f_len=%s" % (rf_index, rf_len, f_index, f_len))
+            if field in reflection.swapped:
+                debug("field \"%s\" swapped (back)" % (field))
+                reflection.swapped.remove(field)
+                back_swapped.append(field)
+                assert field in reflection.fields, "Reflection for %s indicates swapping %s but swapped position is not indicated in the reflection fields. Should it be ignored?" % (reflection_name, field)
+                assert reflect_field != field, "Reflection for %s should not indicate swapping %s since it is in the correct order" % (reflection_name, field)
+                f_index += 1
+                continue
+            if reflect_field in reflection.swapped:
+                debug("field \"%s\" swapped (fwd)" % (field))
+                reflection.swapped.remove(reflect_field)
+                fwd_swapped.append(reflect_field)
+                assert reflect_field in reflection.fields, "Reflection for %s indicates swapping field %s but it doesn't exist in that class/struct so it should be removed" % (reflection_name, reflect_field)
+                rf_index += 1
+                continue
+            assert reflect_field not in ignored, "Reflection for %s should not indicate %s for %s, it should indicate %s - %s" % (reflection_name, ignore_str, reflect_field, swap_str, ",".join(ignored))
```

Comment:
Semicolon or period in the message, not comma.

Reply:
Ok